Mathlib/LinearAlgebra/Dimension/RankDecomposition.lean

```diff
@@ -0,0 +1,214 @@
+  apply LinearEquiv.ofBijective g
+  constructor
+  · simpa [← ker_eq_bot, g, ker_codRestrict] using f.ker_complement_restriction_injective
+  intro ⟨_, y, hyx⟩
+  use (f.prodEquivOfkerComplement.2 y).1
+  simp [g, codRestrict, ← hyx, ker_complement_restriction, apply_ker_component_eq_zero]
+
+lemma finrank_kerComplement_eq_rank {r : ℕ} (hr : rank f = r) : finrank K f.kerComplement = r := by
+  simp [finrank_eq_of_rank_eq hr, LinearEquiv.finrank_eq f.kerComplementEquivRange]
+
+lemma finrank_ker_eq [FiniteDimensional K V] {r n : ℕ} (hr : rank f = r) (hn : finrank K V = n) :
+    finrank K (ker f) = n - r := by
+  rw [← hn, ← finrank_range_add_finrank_ker f, ← finrank_eq_of_rank_eq hr, add_tsub_cancel_left]
+
+lemma card_cokernel_basis_index_eq {m r : ℕ} [FiniteDimensional K V] [FiniteDimensional K W]
+    (hm : finrank K W = m) (hr : f.rank = r) :
+    Fintype.card (sumExtendIndex f.linear_independent_ker_complement_basis_image) = m - r := by
+  have := finrank_eq_card_basis f.range_decomposition_basis
+  rw [Fintype.card_sum, hm] at this
+  simp [this, ← finrank_eq_card_basis (ofVectorSpace K f.kerComplement),
+    f.finrank_kerComplement_eq_rank hr]
+
+lemma apply_kerComplement_basis_eq_range_basis (j) :
+    f (f.decomposition_basis (Sum.inl j)) = (f.range_decomposition_basis (Sum.inl j)) := by
+  simp [decomposition_basis, prodEquivOfkerComplement, range_decomposition_basis, sumExtend,
+    Equiv.sumCongr, ker_complement_basis_image, ker_complement_restriction]
+
+end LinearMap
+
+end
+
+
+
+section
+
+variable {R : Type} [Field R] {m n r : ℕ} {M₁ M₂ : Type*}
+variable [AddCommGroup M₁] [Module R M₁] [FiniteDimensional R M₁]
+variable [AddCommGroup M₂] [Module R M₂] [FiniteDimensional R M₂] (f : M₁ →ₗ[R] M₂)
+
+open Matrix LinearMap
+
+theorem exists_basis_for_normal_form_abstract :
+  ∃ (v₁ : Basis ((ofVectorSpaceIndex R f.kerComplement) ⊕ (ofVectorSpaceIndex R (ker f))) R M₁)
+    (v₂ : Basis ((ofVectorSpaceIndex R f.kerComplement) ⊕
+      (sumExtendIndex f.linear_independent_ker_complement_basis_image)) R M₂),
+    LinearMap.toMatrix v₁ v₂ f = fromBlocks 1 0 0 0 := by
+  use f.decomposition_basis, f.range_decomposition_basis
+  funext i j
+  match i, j with
+  | Sum.inl i', Sum.inl j' => simp [toMatrix_apply, f.apply_kerComplement_basis_eq_range_basis j',
+      Finsupp.single, Pi.single, Function.update, Matrix.one_apply]
+  | Sum.inr i', Sum.inr j' => simp [toMatrix_apply, decomposition_basis, prodEquivOfkerComplement]
+  | Sum.inl i', Sum.inr j' => simp [toMatrix_apply, decomposition_basis, prodEquivOfkerComplement]
+  | Sum.inr i', Sum.inl j' => simp [toMatrix_apply, f.apply_kerComplement_basis_eq_range_basis j']
+
+theorem exists_basis_for_normal_form (hn : finrank R M₁ = n) (hm : finrank R M₂ = m)
+    (hr : rank f = r) :
+    ∃ (v₁ : Basis (Fin r ⊕ Fin (n - r)) R M₁) (v₂ : Basis (Fin r ⊕ Fin (m - r)) R M₂),
+    f.toMatrix v₁ v₂ = fromBlocks 1 0 0 0 := by
```

Comment:
I don't think this requires finiteness, and I think you shouldn't use a `Sum` type in the final statement but rather say the indexing type of any provided basis.

Reply:
Also you should do this for any complementary submodule to ker f, not just the arbitrarily chosen one.